src/CodeStyle/Core/CodeFixes/FormattingCodeFixProvider.cs

```diff
@@ -1,55 +1,86 @@
 
 using System.Collections.Immutable;
-using System.Composition;
+using System.IO;
 using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.CodeActions;
 using Microsoft.CodeAnalysis.CodeFixes;
 using Microsoft.CodeAnalysis.Diagnostics;
 using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.VisualStudio.CodingConventions;
 
 namespace Microsoft.CodeAnalysis.CodeStyle
 {
-    [ExportCodeFixProvider(LanguageNames.CSharp, LanguageNames.VisualBasic, Name = PredefinedCodeFixProviderNames.FixFormatting)]
-    [Shared]
-    internal class FormattingCodeFixProvider : CodeFixProvider
+    internal abstract class AbstractFormattingCodeFixProvider : CodeFixProvider
     {
-        public override ImmutableArray<string> FixableDiagnosticIds
+        public sealed override ImmutableArray<string> FixableDiagnosticIds
             => ImmutableArray.Create(IDEDiagnosticIds.FormattingDiagnosticId);
 
-        public override FixAllProvider GetFixAllProvider()
+        public sealed override FixAllProvider GetFixAllProvider()
         {
-            return new FixAll();
+            return new FixAll(this);
         }
 
-        public override Task RegisterCodeFixesAsync(CodeFixContext context)
+        public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
         {
             foreach (var diagnostic in context.Diagnostics)
             {
                 context.RegisterCodeFix(
                     CodeAction.Create(
                         CodeStyleResources.Fix_formatting,
-                        c => FormattingCodeFixHelper.FixOneAsync(context.Document, diagnostic, c),
-                        nameof(FormattingCodeFixProvider)),
+                        c => FixOneAsync(context, diagnostic, c),
+                        nameof(AbstractFormattingCodeFixProvider)),
                     diagnostic);
             }
 
             return Task.CompletedTask;
         }
 
+        protected abstract OptionSet ApplyFormattingOptions(OptionSet optionSet, ICodingConventionContext codingConventionContext);
+
+        private async Task<Document> FixOneAsync(CodeFixContext context, Diagnostic diagnostic, CancellationToken cancellationToken)
+        {
+            var options = await GetOptionsAsync(context.Document, cancellationToken).ConfigureAwait(false);
+            return await FormattingCodeFixHelper.FixOneAsync(context.Document, options, diagnostic, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task<OptionSet> GetOptionsAsync(Document document, CancellationToken cancellationToken)
+        {
+            OptionSet options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+
+            // The in-IDE workspace supports .editorconfig without special handling. However, the AdhocWorkspace used
```

Comment:
Is this a test-only code? If so, can it be refactored into a test-only method?

Reply:
It's not strictly test-only code. It's addressing a limitation of the workspace which would naturally go away once we switch the code to using the new .editorconfig support in the compiler.